irohad/execution/impl/common_executor.cpp

```diff
@@ -27,37 +27,34 @@ namespace iroha {
   boost::optional<shared_model::interface::RolePermissionSet>
   getAccountPermissions(const std::string &account_id,
                         ametsuchi::WsvQuery &queries) {
-    auto roles = queries.getAccountRoles(account_id);
-    if (not roles) {
-      return boost::none;
-    }
-    auto r = roles.value();
-    shared_model::interface::RolePermissionSet permissions{};
-    std::for_each(r.begin(), r.end(), [&permissions, &queries](auto &role) {
-      auto perms = queries.getRolePermissions(role);
-      if (not perms) {
-        return;
-      }
-      permissions |= *perms;
-    });
-    return permissions;
+    return queries.getAccountRoles(account_id) | [&queries](const auto &roles) {
+      return std::accumulate(roles.begin(),
+                             roles.end(),
+                             shared_model::interface::RolePermissionSet{},
+                             [&queries](auto &&permissions, const auto &role) {
+                               queries.getRolePermissions(role) |
+                                   [&](const auto &perms) {
+                                     permissions |= perms;
+                                   };
+                               return permissions;
```

Comment:
won't it make a copy?

Reply:
yeah, would be nice to add std::forward

Reply:
Then decltype auto should be added as well